### PR TITLE
Add in checkpointing capability

### PR DESCRIPTION
**Pull request by muellerzr**
Closes https://github.com/huggingface/accelerate/issues/171

This PR adds in two functions, `Accelerator.save_state` and `Accelerator.load_state`, which will go through and checkpoint the current state of everything accelerator touches.

I've included a testing script, but this needs to be properly turned into a test, as currently it is just a script with a bunch of asserts. I have tested successfully on:

- Single GPU
- Single CPU
- Multi-GPU

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -560,6 +566,109 @@ def save(self, obj, f):
         """
         save(obj, f)
 
+    def save_state(self, output_dir: str):
```
Comment (sgugger) [inline src/accelerate/accelerator.py:569]: As much as possible, I would prefer to have everything in another module and have `accelerator` be a thin wrapper. I think we could have a new `checkpointing.py` file where a function that does all of this (taking `models`, `optimizers`, `scaler`). Same for the loading.

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -560,6 +566,109 @@ def save(self, obj, f):
...
+        output_dir = os.path.expanduser(output_dir)
+        os.makedirs(output_dir, exist_ok=True)
+        logger.info(f"Saving current state to {output_dir}")
+
+        # Model states
+        for i, model in enumerate(self._models):
+            state = self.get_state_dict(model)
+            weights_name = "pytorch_model"
+            if i == 0:
+                weights_name += f"_{i}"
```
Comment (sgugger) [inline src/accelerate/accelerator.py:587]:
```suggestion
            if i != 0:
                weights_name += f"_{i}"
```
Is it a typo? I think it's for every model **but** the first one.

File: src/accelerate/accelerator.py [hunk 3]
```diff
@@ -560,6 +566,109 @@ def save(self, obj, f):
...
+            weights_name += ".bin"
+            output_model_file = os.path.join(output_dir, weights_name)
+            torch.save(state, output_model_file)
```
Comment (sgugger) [inline src/accelerate/accelerator.py:590]: Should be `accelerator.save` here (otherwise won't work for TPUs).

File: src/accelerate/accelerator.py [hunk 4]
```diff
@@ -560,6 +566,109 @@ def save(self, obj, f):
...
+            logger.info(f"Model weights saved in {output_model_file}")
+        # Optimizer states
+        for i, opt in enumerate(self._optimizers):
+            state = opt.state_dict()
+            optimizer_name = "optimizer"
+            if i == 0:
+                optimizer_name += f"_{i}"
+            optimizer_name += ".pt"
+            output_optimizer_file = os.path.join(output_dir, optimizer_name)
+            torch.save(state, output_optimizer_file)
```
Comment (sgugger) [inline src/accelerate/accelerator.py:600]: Same here:
```suggestion
            accelerator.save(state, output_optimizer_file)
```

File: src/accelerate/accelerator.py [hunk 5]
```diff
@@ -560,6 +566,109 @@ def save(self, obj, f):
...
+            logger.info(f"Optimizer state saved in {output_optimizer_file}")
+        # GradScalar state
+        if self.scaler is not None:
+            state = self.scaler.state_dict()
+            scalar_name = "scaler.bin"
+            output_scalar_file = os.path.join(output_dir, scalar_name)
+            torch.save(state, output_scalar_file)
+            logger.info(f"GradScalar state saved in {output_scalar_file}")
```
Comment (sgugger) [inline src/accelerate/accelerator.py:608]:
```suggestion
            logger.info(f"Gradient scaler state saved in {output_scaler_file}")
```

File: src/accelerate/accelerator.py [hunk 6]
```diff
@@ -560,6 +566,109 @@ def save(self, obj, f):
...
+            optimizer_name += ".pt"
+            output_optimizer_file = os.path.join(output_dir, optimizer_name)
+            torch.save(state, output_optimizer_file)
+            logger.info(f"Optimizer state saved in {output_optimizer_file}")
+        # GradScalar state
+        if self.scaler is not None:
+            state = self.scaler.state_dict()
+            scalar_name = "scaler.bin"
+            output_scalar_file = os.path.join(output_dir, scalar_name)
+            torch.save(state, output_scalar_file)
```
Comment (sgugger) [inline src/accelerate/accelerator.py:607]:
```suggestion
            output_scaler_file = os.path.join(output_dir, scaler_name)
            torch.save(state, output_scaler_file)
```

File: src/accelerate/accelerator.py [hunk 7]
```diff
@@ -560,6 +566,109 @@ def save(self, obj, f):
...
+        for i, opt in enumerate(self._optimizers):
+            state = opt.state_dict()
+            optimizer_name = "optimizer"
+            if i == 0:
+                optimizer_name += f"_{i}"
+            optimizer_name += ".pt"
+            output_optimizer_file = os.path.join(output_dir, optimizer_name)
+            torch.save(state, output_optimizer_file)
+            logger.info(f"Optimizer state saved in {output_optimizer_file}")
+        # GradScalar state
```
Comment (sgugger) [inline src/accelerate/accelerator.py:602]: I realize we are missing the learning rate scheduler -> we should thus add it as an optional argument to the `save_state` function, or add a `register_for_checkpoints` method to the Accelerator object that would take objects not passed to the `prepare` method but the user still wants to save.

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this! I've left a few comments on the general API and some nits. The most important thing is to figure out how to save the LR scheduler with the rest :-)

File: tests/test_state_checkpointing.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        (a3, b3) = model_unwrapped.a.item(), model_unwrapped.b.item()
+        opt_state3 = optimizer.state_dict()
+        assert a1 == a3
+        assert b1 == b3
+        assert opt_state1 == opt_state3
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Simple example of training script.")
+    parser.add_argument("--fp16", action="store_true", help="If passed, will use FP16 training.")
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply tests/test_state_checkpointing.py]: I'd like some advice on how to convert my script -> a proper test. I tried looking in `test_multigpu` but still doesn't quite line up with how the testing workflow operates

File: src/accelerate/accelerator.py [hunk 9]
```diff
@@ -560,6 +562,36 @@ def save(self, obj, f):
         """
         save(obj, f)
 
+    def save_state(self, output_dir: str):
+        """
+        Saves the current states of the model, optimizer, scaler, and RNG generators to `folder_name`.
```
Comment (sgugger) [inline src/accelerate/accelerator.py:567]:
```suggestion
        Saves the current states of the model, optimizer, scaler, and RNG generators.
```
It's actually output_dir but since we document it just after, no need to mention it up there I believe.

File: src/accelerate/accelerator.py [hunk 10]
```diff
@@ -560,6 +562,36 @@ def save(self, obj, f):
...
+        # Check if folder exists
+        output_dir = os.path.expanduser(output_dir)
+        os.makedirs(output_dir, exist_ok=True)
+        logger.info(f"Saving current state to {output_dir}")
+        weights = [self.get_state_dict(m) for m in self._models]
+        return save_accelerator_state(output_dir, weights, self._optimizers, self.state.process_index, self.scaler)
+
+    def load_state(self, input_dir: str):
+        """
+        Loads the current states of the model, optimizer, scaler, and RNG generators from `folder_name`.
```
Comment (sgugger) [inline src/accelerate/accelerator.py:582]:
```suggestion
        Loads the current states of the model, optimizer, scaler, and RNG generators.
```

File: src/accelerate/checkpointing.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
+# Copyright 2021 The HuggingFace Team. All rights reserved.
```
Comment (sgugger) [inline src/accelerate/checkpointing.py:1]:
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

File: src/accelerate/checkpointing.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+
+logger = logging.getLogger(__name__)
+
+
+def save_accelerator_state(
+    output_dir: str, model_states: List[dict], optimizers: list, process_index: int, scaler: GradScaler = None
+):
+    """
+    Saves the current states of the `models`, `optimizers`, scaler, and RNG generators to `output_dir`.
```
Comment (sgugger) [inline src/accelerate/checkpointing.py:40]:
```suggestion
    Saves the current states of the models, optimizers, scaler, and RNG generators to a given directory.
```
Let's leave the argument names for the arg descriptions and be moe "English" in the description (or if you want them, you need to put them in double backticks or add an `:obj:` to have them code-formatted)

File: src/accelerate/checkpointing.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    Args:
+        output_dir (:obj:`str` or :obj:`os.PathLike`):
+            The name of the folder to save all relevant weights and states.
+        model_states (:obj:`list`):
+            A list of model states
+        optimizers (:obj:`list`):
+            A list of optimizer instances
+        process_index (:obj:`int`):
+            The current process index in the Accelerator state
+        scaler (:obj:`GradScaler`), Optional:
```
Comment (sgugger) [inline src/accelerate/checkpointing.py:51]:
```suggestion
        scaler (:obj:`torch.cuda.amp.GradScaler`, `optional`):
```

File: src/accelerate/checkpointing.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            An optional `GradScaler` instance to save
```
Comment (sgugger) [inline src/accelerate/checkpointing.py:52]:
```suggestion
            An optional gradient scaler instance to save
```
(The type is already described)

**Review by sgugger — APPROVED** [review]
Thanks a lot for continuing your work on this. Just left some nits and one last comment on the test you added, then we can merge this!

File: tests/test_state_checkpointing.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        "between fp16 and bf16 (bfloat16). Bf16 requires PyTorch >= 1.10."
+        "and an Nvidia Ampere GPU.",
+    )
+    parser.add_argument("--cpu", action="store_true", help="If passed, will train on the CPU.")
+    args = parser.parse_args()
+    CheckpointTest().test_can_resume_training(args)
+
+
+if __name__ == "__main__":
+    main()
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply tests/test_state_checkpointing.py:144]: @sgugger how would we delegate the tests to be ran on cpu, mixed precision, etc then in that case?
